### ISSUE TITLE
Conditional elimination phase before static analysis breaks debuginfotest gate

### ISSUE THREAD

**Issue by zakkak** [oracle/graal#4036]
**Describe the issue**
As discussed in https://github.com/oracle/graal/issues/4018#issuecomment-973888501 after the merge of https://github.com/oracle/graal/commit/ee638aadb6201aa713e44c4865858a84f856d06e the debuginfotest started failing in Java 17 with:

```
(gdb) info func java.io.PrintStream::println
Checker info func java.io.PrintStream::println: insufficient matching lines 0 for regular expressions 1
Checker info func java.io.PrintStream::println {
  re.compile('[ \t]*void java.io.PrintStream::println\\(java\\.lang\\.String \\*\\)')
}

All functions matching regular expression "java.io.PrintStream::println":

File java/io/PrintStream.java:
	void java.io.PrintStream::println(java.lang.Object *);
```

**Steps to reproduce the issue**

```shell
git clone https://github.com/oracle/graal.git
cd graal
git checkout 4550c69898bf6c0eec11e4965f86b66b25396268
mx --primary-suite-path substratevm --components='Native Image' build
mx --primary-suite-path substratevm --components='Native Image' debuginfotest
```

**Describe GraalVM and your environment:**
 - GraalVM version: **4550c69898bf6c0eec11e4965f86b66b25396268**
 - JDK major version: **17**
 - OS: **Fedora 34**
 - Architecture: **AMD64**

**Comment by zakkak** [comment]
**TLDR:** the gate failure appears to be a false alarm so I plan to change the test. (FYI @fniephaus)

As [mentioned](https://github.com/oracle/graal/issues/4018#issuecomment-973942928) by @adinn:

> I can certainly see how insertion of the IterativeConditionalEliminationPhase might mean that there is no longer an explicit call to println(String) that causes the method to be compiled. Conditional elimination can remove branches which include non-inlined calls. It can also change inlining decisions, causing calls to be inlined. This might be enough to ensure that all calls only exists inlined i.e. the method does not get compiled independently as a top level method. That would explain the observed behaviour. gbd only lists top-level compiled methods under the info command even though it knows that the method exists and contributes to inline code (i.e. to be more precise info lists methods with declaration and definition but not those with just a declaration).

Indeed after https://github.com/oracle/graal/commit/ee638aadb6201aa713e44c4865858a84f856d06e `println(String) no longer appears as a primary range (i.e. top-level method). Note that a Debug Info Entry (DIE) for it is still generated (as expected) when using `-H:-OmitInlinedMethodDebugLineInfo`.

@adinn Do I understand correctly that the reason that `info func` fails to list the methods is the lack of a Frame Description Entry (FDE)?

@olpaw @adinn what do you think the user experience should be in such cases (although this is a different issue that existed even before ee638aadb6201aa713e44c4865858a84f856d06e)?
IMO users should be able to see inlined methods with `info func` and also set breakpoints on them by passing the full name and parameter types, e.g. `b java.io.PrintStream::println(java.lang.String *)` (see https://github.com/oracle/graal/issues/4018#issuecomment-973966978).

**Comment by adinn** [comment]
> @adinn Do I understand correctly that the reason that info func fails to list the methods is the lack of a Frame Description Entry (FDE)?

I don't believe the presence of an FDE  is of any relevance to how info works. An FDE is provided so gdb can understand how to unwind stack frames and identify the associated register save locations. info func does nto need that information to identify which compiled methods exist in the program. As far as I am aware it bases its printout on information it retrieves from the info section. Specifically, it relies on method/function declaration and definition records with tag DW_TAG_subprogram. There are three variants of these (which have the associated listed tags):

  declaration: DW_AT_declaration
  definition: DW_AT_specification, 
  inline definition: DW_AT_specification, DW_AT_inline

Every method known about in the native image should have a declaration

Only methods that occur as top level call targets will have a definition.

Any method that has been inlined will have an inline definition. This may be additional to a definition if the method exists as both inline code and as a top level call target. But in some cases there may only be an inline definition.

The problem is that when there is only an inline definition info func will not show the method. As far as I am aware this is not specific to the Java info model in use by GraalVM. The same thing happens with C programs.

**Comment by zakkak** [comment]
Right, thanks for refreshing my memory (which seems to need an upgrade :sweat_smile:) Andrew!
I indeed see a declaration for `printn(String)`:

```
 <2><1c12739>: Abbrev Number: 21 (DW_TAG_subprogram)
    <1c1273a>   DW_AT_external    : 1
    <1c1273b>   DW_AT_name        : (indirect string, offset: 0xa0be7): println
    <1c1273f>   DW_AT_decl_file   : 1
    <1c12741>   DW_AT_type        : <0x4d>
    <1c12745>   DW_AT_artificial  : 0
    <1c12746>   DW_AT_accessibility: 1	(public)
    <1c12747>   DW_AT_declaration : 1
    <1c12748>   DW_AT_containing_type: <0x1c125a2>
    <1c1274c>   DW_AT_object_pointer: <0x1c12750>
```

an inline definition:
```
 <1><1c16171>: Abbrev Number: 13 (DW_TAG_subprogram)
    <1c16172>   DW_AT_inline      : 1	(inlined)
    <1c16173>   DW_AT_external    : 1
    <1c16174>   DW_AT_specification: <0x1c12739>
```

and no (top level) definition.

> The problem is that when there is only an inline definition info func will not show the method. As far as I am aware this is not specific to the Java info model in use by GraalVM. The same thing happens with C programs.

You are right on this one as well. For the following C/C++ program:

```cpp
#include <stdio.h>

__attribute__((always_inline)) inline
void baz() {
    printf("Hello World!");
}

__attribute__((always_inline)) inline
void bar(int i) {
    baz();
}

void bar(float x) {
    baz();
}

void foo() {
    bar((int) 10);
    bar((float) 10.0);
}

int main(int argc, char *argv[]) {

    foo();

    return 0;
}
```

gdb returns:
```
(gdb) info func main
All functions matching regular expression "main":

File inlining.cpp:
22:	int main(int, char**);
(gdb) info func baz
All functions matching regular expression "baz":
(gdb) info func bar
All functions matching regular expression "bar":

File inlining.cpp:
13:	void bar(float);
(gdb) info func foo
All functions matching regular expression "foo":

File inlining.cpp:
17:	void foo();
```

Note that `bar(int)` is missing like in our case.

What I see **different** is that in the C/C++ case I can run `b bar(int)` and it will work (although tab completion doesn't provide this choice) while in our case doing `b java.io.PrintStream::println(java.lang.String *)` I am getting:

```
Function "java.io.PrintStream::println(java.lang.String *)" not defined.
Make breakpoint pending on future shared library load? (y or [n]) Quit
```

In the C/C++ case the corresponding DIE is:

```
 <1><195>: Abbrev Number: 16 (DW_TAG_subprogram)
    <196>   DW_AT_external    : 1
    <196>   DW_AT_name        : bar
    <19a>   DW_AT_decl_file   : 1
    <19b>   DW_AT_decl_line   : 9
    <19c>   DW_AT_decl_column : 6
    <19d>   DW_AT_linkage_name: (indirect string, offset: 0x55): _Z3bari
    <1a1>   DW_AT_inline      : 3       (declared as inline and inlined)
    <1a2>   DW_AT_sibling     : <0x1b1>
```

So it looks like that if we provide `DW_AT_decl_line` and `DW_AT_decl_column` for inlined methods we will also be able to set breakpoints at them. I will prepare a PR for this.

**Comment by adinn** [comment]
@zakkak It is unfortunate that bytecode does not contain line numbers for method and field declarations. Line numbers are only provided for code lines in the method body that can be associated with some specific bytecode. I was hoping we could eventually remedy this by defining new line number bytecode attributes and getting javac to co-operate in inserting them. However, that's a long haul to achieve and would only help with new JDK releases (backporting bytecode changes is not an option).

That said, I am not actually convinced that it is the lack of line and column numbers that explains the disparity between the Java and C++ examples you cite here. Of course, the way to find out is to debug gdb. I will add that to my list of things to follow up (unless you want to try doing it, in which case feel free and do ask me if you find you need some advice on how to do that and where to look in the gdb code).

For the moment, I think we need to resolve the test failure by assuming that we cannot rely on the println call being inlined -- hence that we have to modify the test to remove the checks affected by this change. The tests using println were just picked on in the original test because they were a simple case to get working. I think we can afford to drop them because we now include other cases where 1) we still check for correct backtraces and breaks and 2) we use @inline/noinline directives to ensure the relevant methods exist top-level only or inline only.

I'd be grateful if you could pick this issue up and get the tests working again as I have a other things to get sorted before I go on PTO next week.

**Comment by zakkak** [comment]
@adinn @fniephaus FYI I am working on it, but as I keep removing/fixing tests new errors are popping out, the last couple of which seem like actual regressions.

**Comment by adinn** [comment]
Ok, this may need to wait until @christianwimmer or @olpaw fix the problem that introduced multiple copies of methods into the native image method list.

**Comment by adinn** [comment]
This issue has not been fixed and should not be closed.

**Comment by zakkak** [comment]
> This issue has not been fixed and should not be closed.

Why do you say so? This issue (breakage due to Conditional Elimination Phase aka https://github.com/oracle/graal/commit/ee638aadb6201aa713e44c4865858a84f856d06e) should indeed be fixed by #4046 , what is still not fixed is #4018 which was introduced with https://github.com/oracle/graal/commit/7b5f83119a8a216a72924aeaa6c116a877cd020e.

**Comment by adinn** [comment]
@zakkak @fniephaus Oh, sorry, I got confused about the issue being closed (I thought this was GR=35118). You are right that this one should be closed.
